### PR TITLE
Patches for Lux integration

### PR DESCRIPTION
**Pull request by avik-pal**
I am pulling out the patches from https://github.com/EnzymeAD/Reactant.jl/pull/5, so that we can get them in faster

### CODE REVIEW

**Review by wsmoses — COMMENTED** [review]
Could array interface be done in an extension package if it's not in base

**Comment by avik-pal** [comment]
> Could array interface be done in an extension package if it's not in base

Done

File: src/Reactant.jl [hunk 1]
```diff
@@ -138,6 +138,13 @@ end
         seen[prev] = res
         return res
     end
+
+    if RT <: TracedRArray
+        @assert typeof(prev) <: RT # prev has concrete shape while RT might not have that
+        res = typeof(prev)(prev.paths, prev.mlir_data)
```
Comment (wsmoses) [inline src/Reactant.jl:144]: This isn't the right way to construct. You should construct it as a stablehlo.constant op (with zeros). I think we already have a similar overload for base.zero. alternatively we should reasonably be able to dispatch to base.zero here

File: src/utils.jl [hunk 2]
```diff
@@ -7,7 +7,7 @@ function transpose_val(val)
     return MLIR.IR.result(MLIR.Dialects.stablehlo.transpose(val, permutation=attr), 1)
 end
 
-function apply(f, args...; kwargs)
+function apply(f, args...; kwargs...)
```
Comment (wsmoses) [inline src/utils.jl:10]: Can you add a test for what needed this. I somehow recall the other way being intentional but maybe not

Comment (avik-pal) [reply src/utils.jl:10]: This came up in computing the gradient of the model with Enzyme. But that code is not completely functional yet, see https://github.com/LuxDL/Lux.jl/pull/665#issuecomment-2133946153

**Comment by wsmoses** [comment]
Ah looking at that error I think we should trace the primal and shadow
arrays into tracedrarrays so then the duplicates invariant holds. We'll
need the data as an rarray anyways to run it

**Comment by wsmoses** [comment]
However separate from this fix (which ought be done regardless) it is
surprising that the original model is not an rarray as I would expect. That
should probably be fixed in lux

File: src/Reactant.jl (hunk 1 above)
Comment (avik-pal) [reply src/Reactant.jl]: Updated to use the code from `Base.fill!`

File: Project.toml [hunk 3]
```diff
@@ -12,16 +12,19 @@ Preferences = "21216c6a-2e73-6563-6e65-726566657250"
...
 ReactantNNlibExt = "NNlib"
 
 [compat]
-Cassette = "0.3"
+ArrayInterface = "7.10"
 CEnum = "0.4, 0.5"
+Cassette = "0.3"
 Enzyme = "0.11, 0.12"
-Reactant_jll = "0.0.6"
+NNlib = "0.9.17"
```
Comment (wsmoses) [inline Project.toml:27]: change this to 0.9, no need for 17 specificity

**Review by wsmoses — APPROVED** [review]

File: Project.toml (hunk 3 above)
Comment (avik-pal) [reply Project.toml]: Done

File: src/overloads.jl [hunk 4]
```diff
@@ -549,6 +549,12 @@ Cassette.overdub(context::TraceCtx, f::typeof(Base.materialize!), args...) = f(a
 @inline Base.copyto!(dest::TracedRArray, bc::Broadcasted{Nothing}) =
     _copyto!(dest, bc) # Keep it for ArrayConflict
 
+@inline function Base.copyto!(dest::TracedRArray{ElType, Shape, N},
```
Comment (avik-pal) [inline src/overloads.jl:552]: @wsmoses can you confirm that this is the right way to implement copyto!? I am getting https://github.com/LuxDL/Lux.jl/pull/665#issuecomment-2134147847 so it is possible I got something wrong

Comment (wsmoses) [reply src/overloads.jl:552]: Yup that is perfect!

**Review by wsmoses — APPROVED** [review]

File: .vscode/settings.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (wsmoses) [inline .vscode/settings.json:1]: remove this and add to .gitignore?

**Pull request merged**